File: 프로그래머스/Level 2/방문 길이.swift
// https://school.programmers.co.kr/learn/courses/30/lessons/49994
import Foundation

func solution(_ dirs:String) -> Int {
    var answer:Int = 0
    var visited:Set<[Int]> = []
    var cx = 0
    var cy = 0
    
    for dir in dirs {
        if dir == "U" && cx < 5 {
            visited.insert([cx, cy]+[cx+1, cy])
            cx += 1
        } else if dir == "D" && cx > -5 {
            visited.insert([cx-1, cy]+[cx, cy])
            cx -= 1
        } else if dir == "R" && cy < 5 {
            visited.insert([cx, cy]+[cx, cy+1])
            cy += 1
        } else if dir == "L" && cy > -5 {
            visited.insert([cx, cy-1]+[cx, cy])
            cy -= 1
        }
    }
    
    //print(visited)
    answer = visited.count
    
    return answer
}

File: 프로그래머스/Level 2/방문 길이.cpp
// https://school.programmers.co.kr/learn/courses/30/lessons/49994
#include <iostream>
#include <string>
#include <map>

using namespace std;

map<pair<pair<int, int>, pair<int, int>>, bool> visited;

int solution(string dirs) {
    int answer = 0;
    int cx = 5, cy = 5;
    
    for (int i = 0; i < dirs.size(); ++i) {
        int sx = cx, sy = cy;
        if (dirs[i] == 'U') {
            if (cx - 1 >= 0) cx--;
            else continue;
        } else if (dirs[i] == 'D') {
            if (cx + 1 <= 10) cx++;
            else continue;
        } else if (dirs[i] == 'R') {
            if (cy + 1 <= 10) cy++;
            else continue;
        } else {
            if (cy - 1 >= 0) cy--;
            else continue;            
        }
        if (visited[{{sx, sy}, {cx, cy}}] == true) continue;
        
        visited[{{sx, sy}, {cx, cy}}] = true;
        visited[{{cx, cy}, {sx, sy}}] = true;
        answer++;
    }
    return answer;
}